### PR TITLE
Milestone5 feature branch

### PR DESCRIPTION
**Pull request by SunnySparks**
In this pull request, I:

-Created the correct associations for making friendships in social media app by following best practices.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -3,10 +3,11 @@ class UsersController < ApplicationController
 
   def index
     @users = User.all
+    @user = User.find_by(params[:user_id])
```
Comment (Elbie-em) [inline app/controllers/users_controller.rb]: Remove this search for a user instance, it is not necessary for the index action.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -3,10 +3,11 @@ class UsersController < ApplicationController
 
   def index
     @users = User.all
```
Comment (Elbie-em) [inline app/controllers/users_controller.rb]: This query makes your view include the current user that is logged in along with other users. Create a query that shows all users excluding the current user.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -3,10 +3,11 @@ class UsersController < ApplicationController
...
+    @user = User.find_by(params[:user_id])
   end
 
   def show
-    @user = User.find(params[:id])
+    @user = User.find_by(params[:user_id])
     @posts = @user.posts.ordered_by_most_recent
```
Comment (Elbie-em) [inline app/controllers/users_controller.rb:10]: This show method should take the correct params passed from the frontend, you are passing ":user_id" that is not existent in the params. It should be params[:id] instead.

File: app/models/user.rb [hunk 4]
```diff
@@ -9,4 +9,30 @@ class User < ApplicationRecord
   has_many :posts
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships
+  has_many :friends, through: :friendships, source: :friend
+
```
Comment (Elbie-em) [inline app/models/user.rb:14]: This association "has_many :friends" has no reference to any model or alias. Kindly fix this.
*Note: Your current friendship model pretty much covers the requesting friends functionality*
If you want to enforce a friends model, kindly do it correctly.

File: app/models/user.rb [hunk 5]
```diff
@@ -9,4 +9,30 @@ class User < ApplicationRecord
...
+  def confirm_friend(user)
+    friendship = inverse_friendships.find { |friendship| friendship.user == user }
+    friendship.confirmed = true
+    friendship.save
+  end
+
+  def friend?(user)
+    friends.include?(user)
+  end
 end
```
Comment (Elbie-em) [inline app/models/user.rb:34]: The methods created above I see have to work around dealing with the creating, accepting, rejecting, and viewing requests, however, I don't see anywhere in the application they have been used, I additionally see a calling of "inverse_friendships" and it is not clear where it is being referenced from.

File: app/views/friendships/_friendships.html.erb [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+<% @friendships.each do |friendship| %>
+  <tr>
+  <%= friendship %>
+  </tr>
+<% end %>
+
+<small>
+  <%= link_to 'Accept friend', friendships_path(friendships), method: 'put', class: 'btn-success p-1 rounded no-link' %>
+  <%= link_to 'Reject request', friendships_path(friendships), method: 'delete', class: 'btn-danger p-1 rounded no-link' %>
+</small>
```
Comment (Elbie-em) [inline app/views/friendships/_friendships.html.erb]: This piece of code should be inside the loop as you want to show this for every request that is sent to the user.

File: app/views/friendships/_friendships.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<div class="border p-3 mb-3 rounded">
+<% @friendships.each do |friendship| %>
+  <tr>
```
Comment (Elbie-em) [inline app/views/friendships/_friendships.html.erb:4]: This is a partial and you are calling a loop to an instance variable, kindly fix this.

File: config/routes.rb [hunk 8]
```diff
@@ -3,12 +3,15 @@
   root 'posts#index'
 
   devise_for :users
+  devise_for :friendships
```
Comment (Elbie-em) [inline config/routes.rb]: Devise should be used for creating a user resource, you are calling to create an auth resource for the friendships model, this is incorrect, kindly nest the friendship resource in the correct resource.

File: app/views/users/_user.html.erb [hunk 9]
```diff
@@ -2,5 +2,6 @@
   Name: <%= user.name %>
   <span class="profile-link">
     <%= link_to 'See Profile',  user_path(user), class: 'profile-link' %>
+    <%= link_to 'Add Friend', friendships_path(friendship: {friend_id: @user.id, user_id: current_user.id}), method: :post%>
   </span>
```
Comment (Elbie-em) [inline app/views/users/_user.html.erb:10]: This post request is referencing a user instance within a partial. you should only reference the friend being user.id as params as your controller already references the current user as user_id when creating a request.

File: app/controllers/friendships_controller.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
+class FriendshipsController < ApplicationController
+  def index
+    @user = User.find_by(params[:user_id])
```
Comment (Elbie-em) [inline app/controllers/friendships_controller.rb]: This user_id param causes your application to always find the first user as it does not exist. Kindly fix this by searching for the current_user with current_user.id

File: app/controllers/friendships_controller.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    @friendships = @user.friendships
+  end
```
Comment (Elbie-em) [inline app/controllers/friendships_controller.rb:5]: This is a requests page and should load all requests sent to this user, kindly fix this by querying the database for friendships that have the current user's id and the friend_id.

File: app/controllers/friendships_controller.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    @friendship = Friendship.find_by(user_id: params[:id], friend_id: current_user.id)
+    @friendship.status = 'accepted'
+
+    if @friendship.save
+      redirect_to root_path, notice: 'Friend Request Sent'
+    else
+      redirect_to root_path, alert: 'Request Already sent'
+    end
+  end
+end
```
Comment (Elbie-em) [inline app/controllers/friendships_controller.rb:44]: Your accept method causes your application to break as it has references to methods that do not exist.
".status is not a field in the friendships table". Kindly fix this logic

File: app/controllers/friendships_controller.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    @user = User.find_by(params[:user_id])
+    @friendship = @user.friendships.find params[:id]
+    @friendship.accept
+  end
+
+  def destroy
+    @user = User.find_by(params[:user_id])
+    @friendship = @user.friendships.find params[:id]
+    @friendship.decline
+  end
```
Comment (Elbie-em) [inline app/controllers/friendships_controller.rb:17]: The "decline" method is not present anywhere in your application, kindly enforce the destroy logic correctly.

File: app/controllers/friendships_controller.rb [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+  def new
+    @friendship = Friendship.new
+  end
+
+  def create
+    @user = User.find_by(params[:user_id])
+    @friendship = Friendship.new(user_id: current_user.id, friend_id: params[:user_id], confirmed: false)
+    @friendship.save!
```
Comment (Elbie-em) [inline app/controllers/friendships_controller.rb:26]: When creating a friendship, your arguments should correspond to your params, in this case, friend_id should reference params[:friend_id] and not user_id which is unexistent.

File: app/controllers/friendships_controller.rb [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    @user = User.find_by(params[:user_id])
+    @friendships = @user.friendships
+  end
+
+  def update
+    @user = User.find_by(params[:user_id])
+    @friendship = @user.friendships.find params[:id]
+    @friendship.accept
+  end
+
```
Comment (Elbie-em) [inline app/controllers/friendships_controller.rb:12]: From your logic and inspection of your model, you should create the inverse friendship from here. You can generate a method on your model to do this and create a call back that creates the inverse friendship. 
What this means is once confirmed for a friendship is "true", create another friendship by reversing the id's for user and friend and both entries should be true.
e.g user_id: 1 friend_id: 2 confirmed: true <=====> user_id: 2 friend_id:1 confirmed: true.

File: app/controllers/friendships_controller.rb (hunk 12 above)
Comment (Elbie-em) [reply app/controllers/friendships_controller.rb:44]: There is a continuous incorrect reference for the current_user id as user_id, kindly fix this in all places I have indicated.

File: app/views/friendships/_friendships.html.erb [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+  <%= friendship %>
+  </tr>
```
Comment (Elbie-em) [inline app/views/friendships/_friendships.html.erb:8]: When this is corrected, your iteration should be a ruby object, kindly search for the user who has sent this request and display their name instead.

**Review by Elbie-em — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED 🟡  ♻️ 

Hello @SunnySparks, you have made a great effort so far. The design of your friendships model is good to tackle the task at hand, however, there are a number of fixes you need to apply before this milestone is approved. Kindly check the comments below ⬇️ 

## [OPTIONAL]

- [x] In addition to all comments, kindly add a link to direct users to friendships.
- [x] Add some logic that will not allow users to send another request if It already exists(you could not show Add Friend).


Request another code review once you have applied the requested changes.

*Feel free to contact me if you have any questions, All the best 💪 *


> Elbie Moonga (Part-time TSE)

> Github (@Elbie-em)
> Slack (Elbie Moonga)
> Twitter ([@ElbieEm](https://twitter.com/ElbieEm))
> LinkedIn([Elbie Moonga](https://www.linkedin.com/in/elbiemoonga/))

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUIRED** :recycle: :recycle: 

@SunnySparks , you're making decent effort. :+1: 
:heavy_check_mark: Your repo looks professional without linter errors and you've also worked with the correct gitflow.

**To improve** :hammer_and_pick: :construction_worker: 

- [x] A lot of the required changes from the previous review have not been implemented. I encourage you to re-visit the comments and tick them off one at a time. Please don't be frustrated, you're almost there. 

This is the reason your friendship feature is not functional. Currently, two users could send friendship request to each other countless times. Friendship can not be confirmed nor declined.

Please, if you need help, ask for it early. 

[OPTIONAL]

- [ ] I saw just three test examples. Please write tests for associations, validations and integration.

Please submit another CRR when you're ready.

Cheers :smile: With respect :tophat:

**Review by carlos-ssh — APPROVED** [review]
Hola, @SunnySparks:  

 ¡Tu proyecto está completo! No hay nada más que decir aparte de ... es hora de fusionarlo: shipit:  
 ¡Felicidades! 🎉  



 ¡Salud y codificación feliz! 👏👏👏 

 No dude en dejar cualquier pregunta o comentario en el hilo de relaciones públicas si algo no está 100% claro. 


 ------ 
 _ Como se describe en la [Política de límites de revisiones de código] (https://microverse.zendesk.com/hc/en-us/articles/1500004088561), solo tiene 3 revisiones de código por proyecto. Si cree que la revisión del código no fue justa, puede solicitar una segunda opinión utilizando [este formulario] (https://airtable.com/shrQAqnBwek5a0O0s) ._

**Pull request merged**